WP_Query ID caching follow up

peterwilsoncc opened the pull request:
https://core.trac.wordpress.org/ticket/22176

cc @spacedmonkey, @tillkruss -- still a work in progress in case something else comes in but making sure you have eyes on it.

peterwilsoncc committed:
Improve detection of taxonomy queries when generating cache key.

peterwilsoncc committed:
Want to make sure the query count is NOT the same.

peterwilsoncc commented on src/wp-includes/class-wp-query.php:
```diff
@@ -3094,7 +3094,7 @@ public function get_posts() {
 
 			$last_changed = wp_cache_get_last_changed( 'posts' );
-			if ( ! empty( $this->tax_query->queried_terms ) ) {
+			if ( ! empty( $this->tax_query->queries ) ) {
```
I considered testing for any SQL clauses as an alternative here:

```php
if ( isset( $this->tax_query ) ) {
	$tax_query_sql = $this->tax_query->get_sql( $wpdb->posts, 'ID' );
	if (
		! empty( $tax_query_sql['join'] ) ||
		! empty( $tax_query_sql['where'] )
	) {
		$last_changed .= wp_cache_get_last_changed( 'terms' );
	}
}
```

peterwilsoncc committed:
Coding standards fixes.

peterwilsoncc committed:
More CS.

At least I’ve figured out which of the phpcs extensions work.

peterwilsoncc committed:
Test removing a term from a post.

peterwilsoncc committed:
Consistent message format for new tests.

costdev commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -893,4 +893,94 @@ public function test_query_cache_delete_term() {
 	}
+
+	/**
+	 * @ticket 22176
```
Is there an appropriate `@covers` tag for these tests? If not, `@coversNothing` should be used.

costdev commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -893,4 +893,94 @@ public function test_query_cache_delete_term() {
+	/**
+	 * @ticket 22176
+	 */
+	public function test_query_cache_exclude_term() {
```
The test names don't describe the expected result. This means reviewing the contents of the test methods to try to ascertain what's being tested and what's expected, and can lead to misunderstandings.

I'd suggest using:
- `test_query_cache_should_exclude_post_with_excluded_term()`.
- `test_query_cache_should_exclude_post_when_excluded_term_is_added_after_caching()`.
- `test_query_cache_should_not_exclude_post_when_excluded_term_is_removed_after_caching()`.

P.S. If any of these test method names are incorrect, that may prove my initial point 😝

costdev commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -893,4 +893,94 @@ public function test_query_cache_delete_term() {
+			),
+		);
+
+		$post_ids_q1 = get_posts( $args );
```
This and `$post_ids_q2` are just named `$post_ids` in the other two test methods. I think it would be beneficial to keep the naming scheme consistent. By using `$post_ids_q1/$post_ids_q2`, a later reference to `$post_ids` isn't confused with the first reference.

costdev reviewed (changes requested):
Thanks for putting the PR together Peter!

Just a few suggestions in this review.

peterwilsoncc committed:
Merge branch 'trunk' into 22176-wp-query-cache-follow-up

peterwilsoncc committed:
Add `@covers` to test suite.

@covers

peterwilsoncc committed:
Modify names of terms.

peterwilsoncc committed:
Rename variables to indicate which query they belong to.

peterwilsoncc commented:
@costdev 

* In 06f2e1bd74c55da6a6afa140bfec39fa35cd33d2 I've added `@covers WP_Query::get_posts` at the class level
* In fc62eae67c5f98fdcf490cc6c3a9b4413ed477a4 I've renamed the methods
* In d0a7e8bea8eb5f21aa6340037f668ca19a6b39c1 I've modified the variable names to indicate which query the posts belong to

costdev reviewed (approved):
LGTM :+1:

peterwilsoncc commented:
Merged in https://core.trac.wordpress.org/changeset/54111 / 48f17267fb
